drivers/ethernet/eth_stellaris.c

```diff
@@ -0,0 +1,378 @@
+		struct net_eth_hdr *pkt_hdr;
+
+		/* Update statistics counters */
+		eth_stats_update_bytes_rx(iface, frame_len - 6);
+		eth_stats_update_pkts_rx(iface);
+		pkt_hdr = NET_ETH_HDR(pkt);
+		if (net_eth_is_addr_broadcast(&pkt_hdr->dst)) {
+			eth_stats_update_broadcast_rx(iface);
+		} else if (net_eth_is_addr_multicast(&pkt_hdr->dst)) {
+			eth_stats_update_multicast_rx(iface);
+		}
+	}
+
+	return;
+
+err_rx:
+	LOG_ERR("Failed to append data to buffer");
+
+pkt_unref:
+	net_pkt_unref(pkt);
+
+err_mem:
+	eth_stellaris_rx_error(iface);
+}
+
+static void rx_isr(void *arg)
+{
+	/* Read the interrupt status */
+	struct device *dev = (struct device *)arg;
+	struct eth_stellaris_runtime *dev_data = DEV_DATA(dev);
+	int isr_val = sys_read32(REG_MACRIS);
+	u32_t lock;
+
+	lock = irq_lock();
+
+	/* Acknowledge the interrupt */
+	sys_write32(isr_val, REG_MACRIS);
+
+	if (isr_val & BIT_MACRIS_RXINT) {
+		eth_stellaris_rx(dev);
+	}
+
+	if (isr_val & BIT_MACRIS_TXEMP) {
+		dev_data->tx_err = false;
+		k_sem_give(&dev_data->tx_sem);
+	}
+
+	if (isr_val & BIT_MACRIS_TXER) {
+		LOG_ERR("Txn Frame Error");
+		eth_stats_update_errors_tx(dev_data->iface);
+		dev_data->tx_err = true;
+		k_sem_give(&dev_data->tx_sem);
+	}
+
+	if (isr_val & BIT_MACRIS_RXER) {
+		LOG_ERR("Error Frame Received");
+		eth_stellaris_rx_error(dev_data->iface);
+	}
+
+	irq_unlock(lock);
+}
+
+static void eth_stellaris_init(struct net_if *iface)
+{
+	struct device *dev;
+	struct eth_stellaris_config *dev_conf;
+	struct eth_stellaris_runtime *dev_data;
+
+	dev = net_if_get_device(iface);
+	dev_data = DEV_DATA(dev);
+	dev_data->iface = iface;
+	dev_conf = DEV_CFG(dev);
+
+	/* Assign link local address */
+	net_if_set_link_addr(iface,
+			     dev_data->mac_addr, 6, NET_LINK_ETHERNET);
+
+	ethernet_init(iface);
+
+	/* Initialize semaphore */
+	k_sem_init(&dev_data->tx_sem, 0, 1);
+
+	/* Initialize Interrupts */
+	dev_conf->config_func(dev);
+}
+
+static struct net_stats_eth *eth_stellaris_stats(struct device *dev)
+{
+	return &(DEV_DATA(dev)->stats);
+}
+
+static void eth_stellaris_irq_config(struct device *dev);
+
+struct eth_stellaris_config eth_cfg = {
+	.mac_base = CONFIG_ETH_BASE_ADDR,
+	.config_func = eth_stellaris_irq_config,
+};
+
+struct eth_stellaris_runtime eth_data = {
+	.mac_addr = {
+		(u8_t)CONFIG_ETH_MAC_ADDR_0,
+		(u8_t)CONFIG_ETH_MAC_ADDR_1,
+		(u8_t)CONFIG_ETH_MAC_ADDR_2,
+		(u8_t)CONFIG_ETH_MAC_ADDR_3,
+		(u8_t)CONFIG_ETH_MAC_ADDR_4,
+		(u8_t)CONFIG_ETH_MAC_ADDR_5
+	},
+	.tx_err = false,
+	.tx_word = 0,
+	.tx_pos = 0,
+};
+
+static const struct ethernet_api eth_stellaris_apis = {
+	.iface_api.init	= eth_stellaris_init,
+	.iface_api.send = eth_stellaris_send,
+	.get_stats = eth_stellaris_stats,
+};
+
+NET_DEVICE_INIT(eth_stellaris, CONFIG_ETH_DRV_NAME,
+		eth_stellaris_dev_init, &eth_data, &eth_cfg,
+		CONFIG_ETH_INIT_PRIORITY,
+		&eth_stellaris_apis, ETHERNET_L2,
+		NET_L2_GET_CTX_TYPE(ETHERNET_L2), ETH_MTU);
+
+static void eth_stellaris_irq_config(struct device *dev)
```

Comment:
move that above your eth_cfg struct

Reply:
This is not possible, because eth_stellaris_irq_config() references the eth_stellaris device structure. The device structure is defined by NET_DEVICE_INIT(), and hence the eth_stellaris_irq_config can only appear after NET_DEVICE_INIT().

Reply:
@pfalcon Thanks for the tip, this has been fixed.

Reply:
stm32 is especially complex, but yes that's the way.

I am asking that we stay consistent because when it will be time to move these macros to codegen, it will save quite some time.